### PR TITLE
Last round of suggestions by clippy

### PR DESCRIPTION
**Pull request by crawford**
This is a follow-on to https://github.com/smoltcp-rs/smoltcp/pull/402 and is the last in the series. Now that all of the warnings have been addressed, I've also set `cargo clippy` to fail when it encounters warnings going forward.

Should we pin to a specific toolchain or do folks think it's feasible to keep up-to-date with the latest stable? I don't have a sense for how often new lints are added.

### CODE REVIEW

File: examples/benchmark.rs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+#![allow(clippy::collapsible_if)]
```
Comment (whitequark) [inline examples/benchmark.rs:1]: Should we just disable clippy in examples?

File: src/phy/loopback.rs [hunk 2]
```diff
@@ -14,7 +14,7 @@ use crate::phy::{self, Device, DeviceCapabilities};
 use crate::time::Instant;
 
 /// A loopback device.
-#[derive(Debug)]
+#[derive(Debug, Default)]
```
Comment (whitequark) [inline src/phy/loopback.rs]: I don't think this is desirable; end users should use `Loopback::new`. `Loopback` in this case behaves like an OS resource, not just an object.

**Review by whitequark — CHANGES_REQUESTED** [review]
I've reviewed individual commits and I agree with almost all of your judgement calls (except where noted). Good work!

File: examples/benchmark.rs (hunk 1 above)
Comment (crawford) [reply examples/benchmark.rs:1]: I tend to look at the examples first when I'm looking at a new project, so my _personal_ preference would be to make sure that the examples are just as tidy. But I'm happy to drop this if you'd rather not bother.

**Comment by crawford** [comment]
I want to double check that the clippy check actually fails this time. I've got one final commit to address the last lint.

File: examples/benchmark.rs (hunk 1 above)
Comment (whitequark) [reply examples/benchmark.rs:1]: No objection if you prefer it this way.

**Review by whitequark — APPROVED** [review]

**Comment by crawford** [comment]
Today I learned about [`pulL_request_target`](https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/):

> we’ve added a new pull_request_target event, which behaves in an almost identical way to the pull_request event with the same set of filters and payload. However, instead of running against the workflow and code from the merge commit, the event runs against the workflow and code from the base of the pull request.

That's why the job isn't failing. I'm reasonably confident this will work as-is.

**Pull request merged**